gateway-js/src/index.ts

```diff
@@ -703,7 +702,7 @@ export class ApolloGateway implements GraphQLService {
     };
   }
 
-  public onSchemaChange(callback: SchemaChangeCallback): Unsubscriber {
+  public onSchemaChange(callback: (schema: GraphQLSchema, supergraphSdl: string) => void): Unsubscriber {
```

Comment:
This seems like a backwards-incompatible change (for TypeScript users) where existing Apollo Server won't be able to pass in the function it currently passes in. Though maybe I'm wrong due to how assignability works in TypeScript. Perhaps you could write a tiny test that shows that passing in a single-arg callback still compiles?

Hmm. OK, I played around and you're right that this won't break using old AS with new Gateway. But I think the other way around may be a challenge: passing a two-arg function to an old Gateway that expects one arg will not compile.  And in fact that makes sense: if you use new AS with old Gateway you will have code in your callback that expects to get two `string`s but the second one won't be there.

I suggest that we define this callback type as `(schema: GraphQLSchema, supergraphSdl?: string) => void`, and in apollo-server you check if supergraphSdl is defined and blow up with some error about needing to upgrade `@apollo/gateway` if it isn't there (or do some other behavior that is ok with no supergraphSdl).

Do you intend to eventually migrate the SchemaChangeCallback type? Leaving it around indefinitely with the wrong typing sounds confusing. Not sure exactly what the constraints here are.

Reply:
Ah you allude to this in https://github.com/apollographql/apollo-server/pull/5187

> This is dependent on apollographql/federation#738 . Specifically, when this PR lands, users will need to have access to a version of @apollo/gateway containing the referenced PR, or else their TS presumably won't compile (specifically, I believe they'll observe that ApolloGateway won't implement the interface GraphQLService, although I haven't checked this yet).

I see that that PR updates the SchemaChangeCallback type too.

I think you're right about what the problem will be. Making the second arg optional would help here.  (Or making an entirely new callback type and onSchemaChangeWithSupergraphSdl entry point! I think second arg optional is better though.)

Reply:
I'm down with making the second positional argument in `SchemaChangeCallback` optional, and having the check essentially be a runtime check instead of a compile-time check. We can give better messaging with a runtime check, and we do a fair number of runtime checks already at gateway start so it's not unexpected. I've updated apollographql/apollo-server#5187 to make the argument optional.